chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -385,60 +390,123 @@ out:
 // historicalConfDetails looks up whether a transaction is already included in a
 // block in the active chain and, if so, returns details about the confirmation.
 func (b *BitcoindNotifier) historicalConfDetails(txid *chainhash.Hash,
+	heightHint, currentHeight uint32) (*chainntnfs.TxConfirmation, error) {
+
+	// First, we'll attempt to retrieve the transaction details using the
+	// backend node's transaction index.
+	txConf, err := b.confDetailsFromTxIndex(txid)
+	if err != nil {
+		return nil, err
+	}
+
+	if txConf != nil {
+		return txConf, nil
+	}
+
+	// If the backend node's transaction index is not enabled, then we'll
+	// fall back to manually scanning the chain's blocks, looking for the
+	// block where the transaction was included in.
+	return b.confDetailsManually(txid, heightHint, currentHeight)
+}
+
+// confDetailsFromTxIndex looks up whether a transaction is already included
+// in a block in the active chain by using the backend node's transaction index.
+// If it is found, the confirmation details are returned. Otherwise, an error
+// is not returned.
+func (b *BitcoindNotifier) confDetailsFromTxIndex(txid *chainhash.Hash,
 ) (*chainntnfs.TxConfirmation, error) {
 
-	// If the transaction already has some or all of the confirmations,
-	// then we may be able to dispatch it immediately.
-	// TODO: fall back to scanning blocks if txindex isn't on.
 	tx, err := b.chainConn.GetRawTransactionVerbose(txid)
-	if err != nil || tx == nil || tx.BlockHash == "" {
-		if err == nil {
-			return nil, nil
-		}
-		// Do not return an error if the transaction was not found.
-		if jsonErr, ok := err.(*btcjson.RPCError); ok {
-			if jsonErr.Code == btcjson.ErrRPCNoTxInfo {
-				return nil, nil
-			}
+	if err != nil {
+		// Avoid returning an error if the transaction was not found.
+		jsonErr, ok := err.(*btcjson.RPCError)
+		if !ok || jsonErr.Code != btcjson.ErrRPCNoTxInfo {
+			return nil, fmt.Errorf("unable to query for txid "+
+				"%v: %v", txid, err)
 		}
-		return nil, fmt.Errorf("unable to query for txid(%v): %v", txid, err)
+	}
+
+	// Make sure we actually retrieved a transaction that is included in a
+	// block. Without this, we won't be able to retrieve its confirmation
+	// details.
+	if tx == nil || tx.BlockHash == "" {
+		return nil, nil
 	}
 
 	// As we need to fully populate the returned TxConfirmation struct,
 	// grab the block in which the transaction was confirmed so we can
 	// locate its exact index within the block.
 	blockHash, err := chainhash.NewHashFromStr(tx.BlockHash)
 	if err != nil {
-		return nil, fmt.Errorf("unable to get block hash %v for historical "+
-			"dispatch: %v", tx.BlockHash, err)
+		return nil, err
```

Comment:
Why was the wrapped error (which had more detail) removed?

Reply:
Fixed.